src/Inciter/AMR/refinement.h

```diff
@@ -0,0 +1,779 @@
+
+                // Use nodes to update edges
+                // All added edges will be locked due to containing intermediate points
+                // Split Outer face  edges
+                tet_store->edge_store.split(A, C, AC, Edge_Lock_Case::intermediate);
+                tet_store->edge_store.split(A, B, AB, Edge_Lock_Case::intermediate);
+                tet_store->edge_store.split(B, C, BC, Edge_Lock_Case::intermediate);
+
+                // Connect D to intermediate points
+                tet_store->edge_store.generate(D, AC, Edge_Lock_Case::intermediate);
+                tet_store->edge_store.generate(D, BC, Edge_Lock_Case::intermediate);
+                tet_store->edge_store.generate(D, AB, Edge_Lock_Case::intermediate);
+                // Connect inner edges
+                tet_store->edge_store.generate(AC, BC, Edge_Lock_Case::intermediate);
+                tet_store->edge_store.generate(AC, AB, Edge_Lock_Case::intermediate);
+                tet_store->edge_store.generate(AB, BC, Edge_Lock_Case::intermediate);
+
+                // Make new Tets
+                //  This is just the node opposite the face plus each pair
+                //  of the news nodes, and the old corner
+                //  FIXME: How to find that near corner programatically?
+
+                // Hard coded solution
+                // A AC AB D
+                // AC AB BC D
+                // AC BC C D
+                // AB B BC D
+
+                size_t num_children = 4;
+                child_id_list_t child = generate_child_ids(tet_id, num_children);
+
+                // Outsides
+                tet_store->add(child[0], AB , AC, A, D, Refinement_Case::one_to_four, tet_id);
+                tet_store->add(child[2], BC, C,  AC, D, Refinement_Case::one_to_four, tet_id);
+                tet_store->add(child[3], B, BC,  AB, D, Refinement_Case::one_to_four, tet_id);
+
+                // Center
+                size_t center_id = child[1]; // 1 to preserve Jacobian order
+                tet_store->add(center_id, AB, BC, AC, D, Refinement_Case::one_to_four, tet_id);
+
+                tet_store->add_center(center_id);
+
+                tet_store->deactivate(tet_id);
+
+            }
+
+            /**
+             * @brief Refine a given tet id into 8 children.
+             * NOTE: Does not do any validity checking (currently?)
+             *
+             * @param tet_id Id of tet to refine
+             */
+            void refine_one_to_eight( size_t tet_id ) {
+
+                if (!check_allowed_refinement(tet_id)) return;
+
+                // Split every edge into two
+                // Makes 4 tets out of the old corners and 3 near mid-points
+                // Make 4 out of the midpoints
+
+                // For Tet {ABCD} need to know all (non-repeating) node pairs
+                // {AB} {AC} {AD} {BC} {BD} {CD}
+                // This can either be hard coded, or generated with a 2d loop
```

Comment:
Probably better and simpler to just keep it unrolled as it is. See also https://github.com/quinoacomputing/quinoa/pull/156/files#r131709593.